### ISSUE TITLE
Slow training solution

### ISSUE THREAD

**Issue by yongleex** [ztangent/multimodal-dmm#3]
Hi, your cost function nll(negative log likelihood) and kld (KL divergence) have some training problems (both gradient vanish and gradient explode) .
Your nll:
![image](https://user-images.githubusercontent.com/23428700/65843799-23e6ac00-e366-11e9-97cc-2deb4e190bea.png)
The gradients:
![image](https://user-images.githubusercontent.com/23428700/65843801-28ab6000-e366-11e9-86e2-21f403dc7ec4.png)
![image](https://user-images.githubusercontent.com/23428700/65843819-3cef5d00-e366-11e9-8ab7-8fc2eb338ec0.png)
The gradient explosion seems to be obvious, and my experiment also demonstrated this loss has gradient vanishing problem if the input x far from the mean value.

If we change it without chaning the meaning:
![image](https://user-images.githubusercontent.com/23428700/65843871-74f6a000-e366-11e9-8cce-ebf7c6c68fe0.png)

After changing it, I fount that it can be trained much faster (500/3000 epoches config is very large number for your original setting). If you changing the loss, you can train the model without any annealing trick, it will not explode from the loss section.

Maybe, it helps. Thank you

**Comment by ztangent** [comment]
Thanks for pointing this out! I'd been wondering why training tended to be unstable, especially for longer temporal sequences. Unfortunately I think if I make this change, the loss function can no longer be interpreted as a weighted ELBO, so I won't make a change for now -- I might consider adding it as an option though.

If you can give some mathematical justification for how changing the loss in this way still ends up maximizing the ELBO, I'd be interested! For e.g., is there a way to multiply the ELBO by some positive variable (e.g. sigma^2) across all terms (i.e. both the log-likelihood and the KL-divergence) so that there are no more vanishing or exploding gradients?

**Comment by yongleex** [comment]
Thanks. This is a time series(RNN) VAE-style model and needs a lot of effort to train it. The nll loss is mathematically correct but indeed contributes to the training problem. I have no idea to find the equivalent math relation between nll and mse loss.  I know that you adopted a lot of tricks (GRU transition, gradient clip, anneal, split, corrupted data, avoiding denominator to zero, etc) to make the training possible, thus I don't think there is an easy way to completely solve it only by changing the loss.
